File: Java/Pen.java

public class Pen extends Item{


	


	//variables 

    private String color ;
	
	private double tipSize ;
	
	
	Pen(String name,double price , String description , int stock , int id,String color , double tipSize) throws IllegalCharacterException{
		
		super(name,price , description , stock ,id);  	//calls Item constructor
		
		if (color.contains(";"))
			throw new IllegalCharacterException();
		
		this.color = color ;
		
		this.tipSize = tipSize ;
		
	} 


	
	//methods

//----------------getDetails implementation---------
	
	
	public String getDetails() {
		
		return ", Color: " + this.color + ", Tip: " + Double.toString(tipSize) + " mm";
				
	}
	
	
//---------------Setters-----------------------------
	
	
	public void setColor(String c) throws IllegalCharacterException {    	//color
		
		if (c.contains(";"))
			throw new IllegalCharacterException();
		
		this.color = c ;
		
	}
	
	public void SetTipSize(double size) {		//TipSize
		
		this.tipSize = size ;
		
	}
	

//------------Getters-----------------------------
	
	
	public String getColor() {		//Color
		
		return this.color ;
		
	}
	
	public double getTipSize() {		//TipSize
		
		return this.tipSize ;
		
	}

}

File: C++/Pen.cpp
#include "Pen.h"

Pen::Pen(string n, double p, string d, int s, int i, string c, double t): Item(n, p, d, s, i), color(c), tipSize(t) {

    if (color.find(";") != string::npos)
        throw IllegalCharacterException();

}

string Pen::getDetails() {
    ostringstream detailsStream;
    detailsStream << fixed << setprecision(2) << ", Color: " << this->color << ", Tip: " << this->tipSize << " mm";

    return detailsStream.str();
}

string Pen::getCategory() {
    return "pen";
}

void Pen::setColor(string c) {

    if (c.find(";") != string::npos)
        throw IllegalCharacterException();

    this->color = c;

}

string Pen::getColor() {
    return this->color;
}

void Pen::setTipSize(double tSize) {
    this->tipSize = tSize;
}

double Pen::getTipSize() {
    return this->tipSize;
}
